storage/kafka/src/main/java/io/apicurio/registry/kafka/KafkaRegistryStorage.java

```diff
@@ -48,102 +63,405 @@
+        log.info("Autosnapshot on every {} requests, period: {}, scheduled check: {}",
+                 snapshotRequests, snapshotPeriod, schedulePeriod);
+    }
+
+    @Override
+    public String registryTopic() {
+        return registryTopic;
+    }
+
+    @Override
+    public String snapshotTopic() {
+        return snapshotTopic;
+    }
+
+    @Override
+    public void loadSnapshot(StorageSnapshot snapshot) {
+        storage.putAll(snapshot.getStorage());
+        global.putAll(snapshot.getGlobal());
+        artifactRules.putAll(snapshot.getArtifactRules());
+        globalRules.putAll(snapshot.getGlobalRules());
+    }
+
+    @Override
+    public void start() {
+        executor = new ScheduledThreadPoolExecutor(1);
+        // random delay, so not all nodes start checking at the same time
+        long delay = ThreadLocalRandom.current().nextLong(schedulePeriod * 60);
+        executor.scheduleAtFixedRate(this::check, delay, schedulePeriod, TimeUnit.MINUTES);
+    }
+
+    private void check() {
+        long now = System.currentTimeMillis();
+
+        // first check for any stale CFs -- should not be many
+        Iterator<Map.Entry<UUID, CF<Object>>> iter = cfMap.entrySet().iterator();
+        while (iter.hasNext()) {
+            CF cf = iter.next().getValue();
+            // remove if older then the period we check
+            if (now - cf.getTimestamp() > TimeUnit.MINUTES.toMillis(schedulePeriod)) {
+                iter.remove();
+            }
+        }
+
+        // force snapshot by sending a SNAPSHOT type msg
+        // we should be fine, as there shouldn't be lots of such msgs
+        // plus, we will only apply msg after this snapshot
+        // if it fails, we'll just re-try with snapshot, so no harm
+        if (now - lastSnapshotTime > TimeUnit.MINUTES.toMillis(snapshotPeriod)) {
+            log.info("Forced snapshot: " + get(submitSnapshot(now)));
+        }
+    }
+
+    @Override
+    public void stop() {
+        if (executor != null) {
+            executor.shutdown();
+        }
+    }
+
+    private Reg.RegistryValue.Builder getRVBuilder(Reg.ValueType vt, Reg.ActionType actionType, String artifactId, long version) {
+        Reg.RegistryValue.Builder builder = Reg.RegistryValue.newBuilder()
+                                                             .setVt(vt)
+                                                             .setType(actionType)
+                                                             .setVersion(version);
+
+        if (artifactId != null) {
+            builder.setArtifactId(artifactId);
+        }
+        return builder;
+    }
+
+    private <T> CompletableFuture<T> submitArtifact(Reg.ActionType actionType, String artifactId, long version, ArtifactType artifactType, String content) {
+        Reg.ArtifactValue.Builder builder = Reg.ArtifactValue.newBuilder();
+        if (artifactType != null) {
+            builder.setArtifactType(artifactType.ordinal());
```

Comment:
Is it worth using an explicit `id()`, rather than relying on `ordinal()` which would change if enum values were reordered or removed?

Reply:
Enum should not be changed ... 
@EricWittmann wdyt?

Reply:
I'd probably use the string value if it were me.  But I'm also comfortable with `ordinal()` I think - it's really unlikely that the ordering of the enum will change.  And it's certainly something we control. :)